worker/agent.go

```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
+	}
+	return a.grab()
 }
 
-func (a *agent) grab() {
+func (a *agent) grab() (err error) {
 	outpack := getOutPack()
 	outpack.dataType = rt.PT_GrabJobUniq
-	a.write(outpack)
+	return a.Write(outpack)
 }
 
-func (a *agent) PreSleep() {
-	a.Lock()
-	defer a.Unlock()
+func (a *agent) PreSleep() (err error) {
+	if a.conn == nil {
+		return errors.New("No connection")
+	}
 	outpack := getOutPack()
 	outpack.dataType = rt.PT_PreSleep
-	a.write(outpack)
+	return a.Write(outpack)
 }
 
-func (a *agent) reconnect() error {
-	a.Lock()
-	defer a.Unlock()
-	conn, err := net.Dial(a.net, a.addr)
-	if err != nil {
-		return err
+func (a *agent) grabReconnectState() (success bool) {
+	a.reconnectLock.Lock()
+	defer a.reconnectLock.Unlock()
+	if a.reconnectingActive { // another thread is already reconnecting to server, this thread will exit
+		return false
 	}
-	a.conn = conn
-	a.rw = bufio.NewReadWriter(bufio.NewReader(a.conn),
-		bufio.NewWriter(a.conn))
+	a.reconnectingActive = true // I am first to attempt reconnection
+	return true
+}
+
+// called by owner of reconnect state to tell that it has finished reconnecting
+func (a *agent) resetReconnectState() {
+	a.reconnectingActive = false
+}
+
+func (a *agent) Connect() {
+
+	ownReconnect := a.grabReconnectState()
+
+	if !ownReconnect {
+		//Reconnect collision, this thread will exit and wait on next a.Lock() for other to complete reconnection
+		return
+	}
+	defer a.resetReconnectState() // before releasing client lock we will reset reconnection state
+
+	a.worker.Log(Info, "Trying to Connect to server:", a.addr, "...")
+
+	var conn net.Conn
+	var err error
+
+	for !a.worker.isShuttingDown() {
+		for num_tries := 1; !a.worker.isShuttingDown(); num_tries++ {
+
+			if a.conn != nil {
+				_ = a.conn.Close()
+				a.conn = nil
+			}
+
+			// nil-out the rw pointer since it's no longer valid
+			_ = atomic.SwapPointer((*unsafe.Pointer)(unsafe.Pointer(&a.rw)), nil)
+
+			if num_tries%100 == 0 {
+				a.worker.Log(Info, fmt.Sprintf("Still trying to connect to server %v, attempt# %v ...", a.addr, num_tries))
+			}
+			conn, err = net.Dial(a.net, a.addr)
+			if err != nil {
+				time.Sleep(500 * time.Millisecond)
+				continue
+			}
+
+			break
+		}
 
-	a.worker.reRegisterFuncsForAgent(a)
-	a.grab()
+		if conn == nil {
+			// in case worker is shutting down
+			break
+		}
+		// at this point the server is back online, we will disconnect and reconnect again to make sure that we don't have
+		// one of those dud connections which could happen if we've reconnected to gearman too quickly after it started
+		_ = conn.Close()
+		time.Sleep(3 * time.Second)
+
+		// todo: come up with a more reliable way to determine if we have a working connection to gearman, pehaps by performing a test
+		conn, err = net.Dial(a.net, a.addr)
+		if err != nil {
+			// looks like there is another problem, go back to the main loop
+			time.Sleep(time.Second)
+			continue
+		}
+
+		_ = conn.(*net.TCPConn).SetKeepAlive(true)
```

Comment:
I would really prefer this didn't cast without checking, and and used `net.DialTCP()` instead of `net.Dial()` then casting to a `TCPConn` without first checking the safety of it.

Reply:
seems not having effect on ELBs anyway, removing this.